File: examples/platforms/android/app/src/test/java/com/here/android/test/DefaultsTest.java
/*
 * Copyright (C) 2016-2018 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */
package com.here.android.test;

import static junit.framework.Assert.assertTrue;
import static junit.framework.Assert.assertEquals;

import android.os.Build;
import android.support.compat.BuildConfig;
import com.here.android.RobolectricApplication;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

@RunWith(RobolectricTestRunner.class)
@Config(
  sdk = Build.VERSION_CODES.M,
  application = RobolectricApplication.class,
  constants = BuildConfig.class
)
public final class DefaultsTest {

  private static final Defaults.StructWithDefaults STRUCT_WITH_DEFAULTS =
      new Defaults.StructWithDefaults();

  private static boolean checkIfStructEqualsDefault(final Defaults.StructWithDefaults struct) {

    return struct.intField == 42
        && struct.uintField == 4294967295L
        && struct.floatField == 3.14F
        && struct.boolField == true
        && struct.stringField.equals("some string")
        && struct.enumField == Defaults.SomeEnum.BAR_VALUE;
  }

  @Test
  public void checkJavaDefaults() {

    //act
    Defaults.StructWithDefaults myStruct = new Defaults.StructWithDefaults();

    //assert
    assertTrue(checkIfStructEqualsDefault(myStruct));
  }

  @Test
  public void checkCppDefaults() {

    //act
    Defaults.StructWithDefaults myStruct = Defaults.getDefault();

    //assert
    assertTrue(checkIfStructEqualsDefault(myStruct));
  }

  @Test
  public void checkCppDefaultsImmutable() {

    Defaults.ImmutableStructWithDefaults myStruct = Defaults.getImmutableDefault();

    assertEquals(42, myStruct.intField);
    assertEquals(0, myStruct.uintField);
    assertEquals(3.14F, myStruct.floatField);
    assertEquals(false, myStruct.boolField);
    assertEquals("some string", myStruct.stringField);
    assertEquals(Defaults.SomeEnum.BAR_VALUE, myStruct.enumField);
  }

  @Test
  public void checkJavaSpecialDefaults() {
    Defaults.StructWithSpecialDefaults special = new Defaults.StructWithSpecialDefaults();

    assertTrue(Defaults.isNan(special.floatNanField));
    assertTrue(Defaults.isInfinity(special.floatInfinityField));
    assertTrue(Defaults.isInfinity(special.floatNegativeInfinityField));
    assertTrue(Defaults.isNan(special.doubleNanField));
    assertTrue(Defaults.isInfinity(special.doubleInfinityField));
    assertTrue(Defaults.isInfinity(special.doubleNegativeInfinityField));
  }

  @Test
  public void checkCppSpecialDefaults() {
    Defaults.StructWithSpecialDefaults special = Defaults.createSpecial();

    assertTrue(Float.isNaN(special.floatNanField));
    assertTrue(Float.isInfinite(special.floatInfinityField));
    assertTrue(Float.isInfinite(special.floatNegativeInfinityField));
    assertTrue(special.floatNegativeInfinityField < 0);
    assertTrue(Double.isNaN(special.doubleNanField));
    assertTrue(Double.isInfinite(special.doubleInfinityField));
    assertTrue(Double.isInfinite(special.doubleNegativeInfinityField));
    assertTrue(special.doubleNegativeInfinityField < 0);
  }
}

File: examples/platforms/dart/test/Defaults_test.dart
// -------------------------------------------------------------------------------------------------
// Copyright (C) 2016-2020 HERE Europe B.V.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0
// License-Filename: LICENSE
//
// -------------------------------------------------------------------------------------------------

import "package:test/test.dart";
import "package:hello/test.dart";
import "../test_suite.dart";

final _testSuite = TestSuite("Defaults");

void main() {
  _testSuite.test("Check defaults", () {
    final result = new Defaults_StructWithDefaults.withDefaults();

    expect(result.intField, 42);
    expect(result.uintField, 4294967295);
    expect(result.floatField, 3.14);
    expect(result.boolField, true);
    expect(result.stringField, "some string");
    expect(result.enumField, Defaults_SomeEnum.barValue);
  });
  _testSuite.test("Check defaults immutable", () {
    final result =
        new Defaults_ImmutableStructWithDefaults.withDefaults(77, true);

    expect(result.intField, 42);
    expect(result.uintField, 77);
    expect(result.floatField, 3.14);
    expect(result.boolField, true);
    expect(result.stringField, "some string");
    expect(result.enumField, Defaults_SomeEnum.barValue);
  });
  _testSuite.test("Check special defaults", () {
    final result = new Defaults_StructWithSpecialDefaults.withDefaults();

    expect(result.floatNanField, isNaN);
    expect(result.floatInfinityField, double.infinity);
    expect(result.floatNegativeInfinityField, double.negativeInfinity);
    expect(result.doubleNanField, isNaN);
    expect(result.doubleInfinityField, double.infinity);
    expect(result.doubleNegativeInfinityField, double.negativeInfinity);
  });
  _testSuite.test("Check empty defaults", () {
    final result = new Defaults_StructWithEmptyDefaults.withDefaults();

    expect(result.intsField, isEmpty);
    expect(result.floatsField, isEmpty);
    expect(result.mapField, isEmpty);
    expect(result.structField.stringField, "some string");
    expect(result.setTypeField, isEmpty);
  });
  _testSuite.test("Check initializer defaults", () {
    final result = new Defaults_StructWithInitializerDefaults.withDefaults();

    expect(result.intsField, [4, -2, 42]);
    expect(result.floatsField, [3.14, double.negativeInfinity]);
    expect(result.mapField, {1: "foo", 42: "bar"});
    expect(result.structField.enumField, Defaults_ExternalEnum.oneValue);
    expect(result.setTypeField, {"foo", "bar"});
  });
}
